File: server/data/wiki-edits.ts
import { EventSource } from "eventsource";
import dotenv from "dotenv";

dotenv.config();

const RETRY_DELAY = 5000;
const BASE_URL = process.env.API_BASE_URL || "http://localhost:8080";

async function postEdit(data: any) {
  const response = await fetch(`${BASE_URL}/wiki/edits`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      title: data.title,
      title_url: data.title_url,
      comment: data.comment,
      user: data.user,
      bot: data.bot,
      notify_url: data.notify_url,
      minor: data.minor,
      length_old: data.length?.old,
      length_new: data.length?.new,
      server_url: data.server_url,
    }),
  });

  if (!response.ok) {
    throw new Error(`Failed to post edit: ${response.statusText}`);
  }
  return response.json();
}

async function connectToEventStream(
  url: string,
  attemptCount = 1,
  duration?: number
): Promise<EventSource> {
  try {
    console.log(
      `Connecting to the EventStreams feed (attempt ${attemptCount})...`
    );
    const eventSource = new EventSource(url);

    eventSource.onmessage = async (event: MessageEvent) => {
      let retryCount = 1;
      while (true) {
        try {
          const data: any = JSON.parse(event.data);
          if (
            data.server_name === "en.wikipedia.org" &&
            data.type === "edit" &&
            data.title &&
            !data.title.includes(":")
          ) {
            await postEdit(data);
            console.log(`edit: ${data.title}`);
            break;
          }
          break;
        } catch (error) {
          console.error(
            `Error processing edit (attempt ${retryCount}):`,
            error
          );
          retryCount++;
          await new Promise((resolve) => setTimeout(resolve, RETRY_DELAY));
        }
      }
    };

    eventSource.onerror = async (error: Event) => {
      console.error("EventSource failed:", error);
      eventSource.close();

      if (!duration) {
        console.log(
          `Retrying connection after failure (attempt ${attemptCount + 1})...`
        );
        await new Promise((resolve) => setTimeout(resolve, RETRY_DELAY));
        await connectToEventStream(url, attemptCount + 1);
      }
    };

    return eventSource;
  } catch (error) {
    console.error("Error establishing connection:", error);
    if (!duration) {
      console.log(
        `Retrying connection after error (attempt ${attemptCount + 1})...`
      );
      await new Promise((resolve) => setTimeout(resolve, RETRY_DELAY));
      return await connectToEventStream(url, attemptCount + 1);
    }
    throw error;
  }
}

export async function main(duration?: number): Promise<void> {
  const url = "https://stream.wikimedia.org/v2/stream/recentchange";
  const eventSource = await connectToEventStream(url, 1, duration);

  if (duration) {
    await new Promise((resolve) => setTimeout(resolve, duration));
    eventSource.close();
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: data/workers/wiki-edits.py
import os
import time
import json
import requests
from sseclient import SSEClient

# Load environment variables
from dotenv import load_dotenv
load_dotenv()

RETRY_DELAY = 5  # in seconds
BASE_URL = os.getenv("API_BASE_URL", "https://pulse-production-58ab.up.railway.app/")

def post_edit(data):
    """
    Posts edit data to the API endpoint.
    """
    url = f"{BASE_URL}/wiki/edits"
    payload = {
        "title": data.get("title"),
        "title_url": data.get("title_url"),
        "comment": data.get("comment"),
        "user": data.get("user"),
        "bot": data.get("bot"),
        "notify_url": data.get("notify_url"),
        "minor": data.get("minor"),
        "length_old": data.get("length", {}).get("old"),
        "length_new": data.get("length", {}).get("new"),
        "server_url": data.get("server_url"),
    }
    headers = {"Content-Type": "application/json"}
    
    response = requests.post(url, headers=headers, json=payload)
    if response.status_code != 200:
        raise Exception(f"Failed to post edit: {response.status_code} {response.text}")
    return response.json()

def connect_to_event_stream(url, attempt_count=1, duration=None):
    """
    Connects to the EventStreams feed and processes incoming events.
    """
    while True:
        try:
            print(f"Connecting to the EventStreams feed (attempt {attempt_count})...")
            client = SSEClient(url)
            
            for event in client:
                try:
                    data = json.loads(event.data)
                    if (
                        data.get("server_name") == "en.wikipedia.org" and
                        data.get("type") == "edit" and
                        data.get("title") and
                        ":" not in data.get("title")
                    ):
                        post_edit(data)
                        print(f"edit: {data.get('title')}")
                except Exception as inner_error:
                    print(f"Error processing edit (retrying): {inner_error}")
                    time.sleep(RETRY_DELAY)

        except Exception as outer_error:
            print(f"Error establishing connection: {outer_error}")
            print(f"Retrying connection after {RETRY_DELAY} seconds (attempt {attempt_count + 1})...")
            time.sleep(RETRY_DELAY)
            attempt_count += 1
        
        if duration:
            time.sleep(duration)
            break

def main(duration=None):
    """
    Main entry point for the script.
    """
    print(BASE_URL)
    url = "https://stream.wikimedia.org/v2/stream/recentchange"
    connect_to_event_stream(url, attempt_count=1, duration=duration)

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(e)
        exit(1)

